File: backend/routes/students.py
"""
Routes for roster student management (CSV upload + pagination).
"""

from __future__ import annotations

import csv
from io import BytesIO, StringIO

from flask import Blueprint, jsonify, request, send_file, session
from sqlalchemy.exc import SQLAlchemyError

from backend.models import db
from backend.services import student_service
from backend.services.student_service import RosterStudentRow

students_bp = Blueprint("students", __name__, url_prefix="/api/students")


@students_bp.get("")
def list_students():
    page = request.args.get("page", default=1, type=int)
    page_size = request.args.get("page_size", default=20, type=int)
    search = request.args.get("search", default="", type=str)
    include_deleted = request.args.get("include_deleted", default=False, type=bool)
    class_name = request.args.get("class_name", default=None, type=str)
    sort_by = request.args.get("sort_by", default="created_at", type=str)
    sort_order = request.args.get("sort_order", default="desc", type=str)

    data = student_service.list_students(
        page=page,
        page_size=page_size,
        search=search,
        include_deleted=include_deleted,
        class_name=class_name,
        sort_by=sort_by,
        sort_order=sort_order,
    )
    return jsonify(data)


@students_bp.post("/add")
def add_students():
    """Add students to the roster bank via CSV upload."""
    if "file" not in request.files:
        return jsonify({"error": "Missing file field"}), 400

    file = request.files["file"]
    if not file.filename:
        return jsonify({"error": "Missing filename"}), 400

    try:
        stream = StringIO(file.stream.read().decode("utf-8"))
    except UnicodeDecodeError:
        return jsonify({"error": "Unable to decode file; make sure it is UTF-8"}), 400

    reader = csv.DictReader(stream)
    rows: list[RosterStudentRow] = []

    for csv_row in reader:
        rows.append(
            RosterStudentRow(
                first_name=csv_row.get("first_name", "") or "",
                last_name=csv_row.get("last_name", "") or "",
                email=csv_row.get("email", "") or "",
            )
        )

    try:
        user_id = session.get("user_id")
        summary, errors, upload_history = student_service.add_students_from_csv(
            enumerate(rows, start=2),  # header is line 1
            filename=file.filename,
            user_id=user_id,
        )
    except SQLAlchemyError as e:
        db.session.rollback()
        return jsonify({"error": "Unable to save roster", "details": str(e)}), 500

    return jsonify({
        "upload_id": upload_history.id,
        "action": "add",
        "summary": summary,
        "errors": errors,
        "upload_history": upload_history.to_dict(),
    }), 201


@students_bp.post("/drop")
def drop_students():
    """Drop (soft delete) students from the roster bank via CSV upload."""
    if "file" not in request.files:
        return jsonify({"error": "Missing file field"}), 400

    file = request.files["file"]
    if not file.filename:
        return jsonify({"error": "Missing filename"}), 400

    try:
        stream = StringIO(file.stream.read().decode("utf-8"))
    except UnicodeDecodeError:
        return jsonify({"error": "Unable to decode file; make sure it is UTF-8"}), 400

    reader = csv.DictReader(stream)
    rows: list[RosterStudentRow] = []

    for csv_row in reader:
        rows.append(
            RosterStudentRow(
                first_name=csv_row.get("first_name", "") or "",
                last_name=csv_row.get("last_name", "") or "",
                email=csv_row.get("email", "") or "",
            )
        )

    try:
        user_id = session.get("user_id")
        summary, errors, upload_history = student_service.drop_students_from_csv(
            enumerate(rows, start=2),  # header is line 1
            filename=file.filename,
            user_id=user_id,
        )
    except SQLAlchemyError as e:
        db.session.rollback()
        return jsonify({"error": "Unable to process drop", "details": str(e)}), 500

    return jsonify({
        "upload_id": upload_history.id,
        "action": "drop",
        "summary": summary,
        "errors": errors,
        "upload_history": upload_history.to_dict(),
    }), 200


@students_bp.post("/upload")
def upload_students():
    """Legacy endpoint - kept for backward compatibility, redirects to /add."""
    # For backward compatibility, treat as add operation
    return add_students()


@students_bp.get("/<int:id>")
def get_student(id: int):
    """Get single student by ID."""
    from backend.models import RosterStudent
    
    student = db.session.get(RosterStudent, id)
    if not student or student.deleted_at:
        return jsonify({"error": "Student not found"}), 404
    
    return jsonify(student.to_dict())


@students_bp.post("")
def create_student():
    """Create a single student manually."""
    from backend.models import RosterStudent
    from sqlalchemy.exc import IntegrityError
    
    data = request.get_json(silent=True) or {}
    email = (data.get("email") or "").strip().lower()
    first_name = (data.get("first_name") or "").strip()
    last_name = (data.get("last_name") or "").strip()
    notes = data.get("notes", "").strip() or None
    class_name = data.get("class_name", "").strip() or None
    
    if not email or not first_name or not last_name:
        return jsonify({"error": "Email, first_name, and last_name are required"}), 400
    
    # Check if already exists (not soft-deleted)
    existing = RosterStudent.query.filter_by(email=email, deleted_at=None).first()
    if existing:
        return jsonify({"error": "Student with this email already exists"}), 409
    
    try:
        student = RosterStudent(
            email=email,
            first_name=first_name,
            last_name=last_name,
            notes=notes,
            class_name=class_name,
            last_updated_via="manual",
        )
        db.session.add(student)
        db.session.commit()
        return jsonify(student.to_dict()), 201
    except IntegrityError:
        db.session.rollback()
        return jsonify({"error": "Student with this email already exists"}), 409


@students_bp.patch("/<int:id>")
def update_student(id: int):
    """Partially update a student (for inline editing)."""
    from backend.models import RosterStudent
    from sqlalchemy.exc import IntegrityError
    
    student = db.session.get(RosterStudent, id)
    if not student or student.deleted_at:
        return jsonify({"error": "Student not found"}), 404
    
    data = request.get_json(silent=True) or {}
    
    # Update fields if provided
    if "email" in data:
        new_email = data["email"].strip().lower()
        if new_email != student.email:
            # Check uniqueness
            existing = RosterStudent.query.filter_by(email=new_email, deleted_at=None).first()
            if existing and existing.id != id:
                return jsonify({"error": "Email already in use"}), 409
            student.email = new_email
    
    if "first_name" in data:
        student.first_name = data["first_name"].strip()
    if "last_name" in data:
        student.last_name = data["last_name"].strip()
    if "notes" in data:
        student.notes = data["notes"].strip() or None
    if "class_name" in data:
        student.class_name = data["class_name"].strip() or None
    
    student.last_updated_via = "inline"
    
    try:
        db.session.commit()
        return jsonify(student.to_dict())
    except IntegrityError:
        db.session.rollback()
        return jsonify({"error": "Email already in use"}), 409


@students_bp.delete("/<int:id>")
def delete_student(id: int):
    """Soft delete a student."""
    from backend.models import RosterStudent
    from datetime import datetime
    
    student = db.session.get(RosterStudent, id)
    if not student:
        return jsonify({"error": "Student not found"}), 404
    
    hard = request.args.get("hard", default=False, type=bool)
    
    if hard:
        # Permanent delete (admin only - consider adding permission check)
        db.session.delete(student)
    else:
        # Soft delete
        student.deleted_at = datetime.utcnow()
        student.last_updated_via = "manual"
    
    db.session.commit()
    return "", 204


@students_bp.patch("/<int:id>/restore")
def restore_student(id: int):
    """Restore a soft-deleted student."""
    from backend.models import RosterStudent
    
    student = db.session.get(RosterStudent, id)
    if not student:
        return jsonify({"error": "Student not found"}), 404
    
    if not student.deleted_at:
        return jsonify({"error": "Student is not deleted"}), 400
    
    student.deleted_at = None
    student.last_updated_via = "manual"
    db.session.commit()
    
    return jsonify(student.to_dict())


@students_bp.delete("/bulk")
def bulk_delete_students():
    """Bulk soft delete multiple students."""
    from backend.models import RosterStudent
    from datetime import datetime
    
    data = request.get_json(silent=True) or {}
    student_ids = data.get("student_ids", [])
    hard = data.get("hard", False)
    
    if not student_ids:
        return jsonify({"error": "student_ids array is required"}), 400
    
    deleted_count = 0
    for student_id in student_ids:
        student = db.session.get(RosterStudent, student_id)
        if student:
            if hard:
                db.session.delete(student)
            else:
                student.deleted_at = datetime.utcnow()
                student.last_updated_via = "manual"
            deleted_count += 1
    
    db.session.commit()
    return jsonify({"deleted": deleted_count})


@students_bp.get("/upload-history")
def get_upload_history():
    """Get list of all CSV uploads with pagination."""
    from backend.models import UploadHistory
    
    page = request.args.get("page", default=1, type=int)
    per_page = request.args.get("per_page", default=20, type=int)
    from_date = request.args.get("from_date", default=None, type=str)
    to_date = request.args.get("to_date", default=None, type=str)
    
    query = UploadHistory.query
    
    # Date filtering (if needed later)
    # if from_date:
    #     query = query.filter(UploadHistory.uploaded_at >= from_date)
    # if to_date:
    #     query = query.filter(UploadHistory.uploaded_at <= to_date)
    
    pagination = query.order_by(UploadHistory.uploaded_at.desc()).paginate(
        page=page, per_page=per_page, error_out=False
    )
    
    return jsonify({
        "uploads": [upload.to_dict() for upload in pagination.items],
        "pagination": {
            "page": pagination.page,
            "per_page": pagination.per_page,
            "total": pagination.total,
            "total_pages": pagination.pages,
        }
    })


@students_bp.get("/upload-history/<int:id>")
def get_upload_details(id: int):
    """Get detailed change log for a specific upload."""
    from backend.models import UploadHistory
    
    upload = db.session.get(UploadHistory, id)
    if not upload:
        return jsonify({"error": "Upload not found"}), 404
    
    return jsonify(upload.to_dict())


@students_bp.get("/template")
def download_template():
    csv_content = student_service.template_csv()
    buffer = BytesIO()
    buffer.write(csv_content.encode("utf-8"))
    buffer.seek(0)

    return send_file(
        buffer,
        mimetype="text/csv",
        as_attachment=True,
        download_name="students_template.csv",
    )

File: src/services/students.ts
import { API_BASE } from './api';

export type Student = {
  id: number;
  email: string;
  first_name: string;
  last_name: string;
  created_at: string;
  updated_at: string;
  deleted_at: string | null;
  notes: string | null;
  class_name: string | null;
  last_updated_via: string | null;
  last_upload_id: number | null;
};

export type Paginated<T> = {
  items: T[];
  page: number;
  page_size: number;
  total: number;
  total_pages: number;
};

export type UploadSummary = {
  added?: number;
  restored?: number;
  skipped?: number;
  removed?: number;
  not_found?: number;
  total_processed: number;
};

export type UploadHistory = {
  id: number;
  filename: string;
  uploaded_at: string;
  uploaded_by: number | null;
  action: 'add' | 'drop' | 'sync';
  summary: {
    added: number;
    updated: number;
    removed: number;
    skipped: number;
    restored: number;
    not_found: number;
    total: number;
  };
  changes: Array<{
    type: string;
    email: string;
    first_name: string;
    last_name: string;
    action: string;
  }>;
};

export type UploadResponse = {
  upload_id: number;
  action: string;
  summary: UploadSummary;
  errors: Array<{ line: number; email?: string; reason: string }>;
  upload_history: UploadHistory;
};

export const studentsService = {
  async list(
    page = 1,
    pageSize = 20,
    search = '',
    includeDeleted = false,
    className?: string,
    sortBy = 'created_at',
    sortOrder: 'asc' | 'desc' = 'desc',
  ): Promise<Paginated<Student>> {
    const params = new URLSearchParams({
      page: String(page),
      page_size: String(pageSize),
      sort_by: sortBy,
      sort_order: sortOrder,
    });
    if (search.trim()) params.set('search', search.trim());
    if (includeDeleted) params.set('include_deleted', 'true');
    if (className) params.set('class_name', className);

    const res = await fetch(`${API_BASE}/students?${params.toString()}`);
    if (!res.ok) throw new Error(`Failed to fetch students (${res.status})`);
    return res.json();
  },

  async get(id: number): Promise<Student> {
    const res = await fetch(`${API_BASE}/students/${id}`);
    if (!res.ok) throw new Error(`Failed to fetch student (${res.status})`);
    return res.json();
  },

  async create(data: {
    email: string;
    first_name: string;
    last_name: string;
    notes?: string;
    class_name?: string;
  }): Promise<Student> {
    const res = await fetch(`${API_BASE}/students`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(data),
      credentials: 'include',
    });
    if (!res.ok) {
      const error = await res.json();
      throw new Error(error.error || `Failed to create student (${res.status})`);
    }
    return res.json();
  },

  async update(
    id: number,
    data: Partial<{
      email: string;
      first_name: string;
      last_name: string;
      notes: string | null;
      class_name: string | null;
    }>,
  ): Promise<Student> {
    const res = await fetch(`${API_BASE}/students/${id}`, {
      method: 'PATCH',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(data),
      credentials: 'include',
    });
    if (!res.ok) {
      const error = await res.json();
      throw new Error(error.error || `Failed to update student (${res.status})`);
    }
    return res.json();
  },

  async delete(id: number, hard = false): Promise<void> {
    const params = hard ? '?hard=true' : '';
    const res = await fetch(`${API_BASE}/students/${id}${params}`, {
      method: 'DELETE',
      credentials: 'include',
    });
    if (!res.ok) throw new Error(`Failed to delete student (${res.status})`);
  },

  async restore(id: number): Promise<Student> {
    const res = await fetch(`${API_BASE}/students/${id}/restore`, {
      method: 'PATCH',
      credentials: 'include',
    });
    if (!res.ok) throw new Error(`Failed to restore student (${res.status})`);
    return res.json();
  },

  async bulkDelete(studentIds: number[], hard = false): Promise<{ deleted: number }> {
    const res = await fetch(`${API_BASE}/students/bulk`, {
      method: 'DELETE',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ student_ids: studentIds, hard }),
      credentials: 'include',
    });
    if (!res.ok) throw new Error(`Failed to bulk delete (${res.status})`);
    return res.json();
  },

  async addFromCsv(file: File): Promise<UploadResponse> {
    const formData = new FormData();
    formData.set('file', file);
    const res = await fetch(`${API_BASE}/students/add`, {
      method: 'POST',
      body: formData,
      credentials: 'include',
    });
    if (!res.ok) {
      const error = await res.json();
      throw new Error(error.error || `Upload failed (${res.status})`);
    }
    return res.json();
  },

  async dropFromCsv(file: File): Promise<UploadResponse> {
    const formData = new FormData();
    formData.set('file', file);
    const res = await fetch(`${API_BASE}/students/drop`, {
      method: 'POST',
      body: formData,
      credentials: 'include',
    });
    if (!res.ok) {
      const error = await res.json();
      throw new Error(error.error || `Drop failed (${res.status})`);
    }
    return res.json();
  },

  async getUploadHistory(page = 1, perPage = 20): Promise<{
    uploads: UploadHistory[];
    pagination: {
      page: number;
      per_page: number;
      total: number;
      total_pages: number;
    };
  }> {
    const params = new URLSearchParams({
      page: String(page),
      per_page: String(perPage),
    });
    const res = await fetch(`${API_BASE}/students/upload-history?${params.toString()}`);
    if (!res.ok) throw new Error(`Failed to fetch upload history (${res.status})`);
    return res.json();
  },

  async getUploadDetails(id: number): Promise<UploadHistory> {
    const res = await fetch(`${API_BASE}/students/upload-history/${id}`);
    if (!res.ok) throw new Error(`Failed to fetch upload details (${res.status})`);
    return res.json();
  },

  downloadTemplate(): void {
    window.open(`${API_BASE}/students/template`, '_blank');
  },
};

